File: apps/mobile/native/ios/DriverIdentifier.swift
// DriverIdentifier.swift — on-device face detection + owner matching
//
// Detection: Vision's VNDetectFaceLandmarksRequest (built into iOS, no
// extra dependency). Recognition needs a small embedding model — same
// recommendation as DriverIdentifier.kt: convert
// `dlib_face_recognition_resnet_model_v1` (davisking/dlib, Boost Software
// License 1.0) or MobileFaceNet to Core ML and bundle it in the app. No
// face image or embedding ever leaves the device — matching happens
// entirely locally against DriverProfile.faceEmbedding
// (packages/core/src/driverProfile.ts).

import Foundation
import Vision
import CoreImage

class DriverIdentifier: NSObject {

    typealias EmbeddingCallback = ([Float], Int64) -> Void
    private var onEmbedding: EmbeddingCallback?

    // TODO: load the converted Core ML embedding model here, e.g.
    // let embeddingModel = try? MobileFaceNet(configuration: MLModelConfiguration())
    var embeddingModelName: String?

    func onFaceEmbedding(_ cb: @escaping EmbeddingCallback) { onEmbedding = cb }

    func process(image: CIImage, timestampUs: Int64) {
        let request = VNDetectFaceLandmarksRequest { [weak self] request, error in
            guard let self = self, error == nil,
                  let faces = request.results as? [VNFaceObservation],
                  let largest = faces.max(by: { $0.boundingBox.width * $0.boundingBox.height < $1.boundingBox.width * $1.boundingBox.height })
            else { return }

            // TODO: crop `largest.boundingBox` from `image`, align via
            // landmarks (eyes/nose), run through the embedding model, then:
            // self.onEmbedding?(embeddingFloats, timestampUs)
            _ = largest
        }
        request.revision = VNDetectFaceLandmarksRequestRevision3

        let handler = VNImageRequestHandler(ciImage: image, options: [:])
        try? handler.perform([request])
    }
}

File: apps/mobile/native/android/src/main/java/com/nexgen/dashcam/DriverIdentifier.kt
package com.nexgen.dashcam

// Driver Identification — on-device face detection + owner matching
//
// Detection: Google ML Kit Face Detection (`com.google.mlkit:face-detection`,
// Apache-2.0, free, fully on-device — no network calls). Recognition
// (turning a detected face into a comparable embedding) needs a small
// extra model; recommended is `dlib_face_recognition_resnet_model_v1`
// (davisking/dlib, Boost Software License 1.0 — free for commercial use,
// keep the license notice) converted to TFLite, or Google's MobileFaceNet.
// Neither model's binary weights are vendored in this repo; drop the
// converted .tflite into android/app/src/main/assets/ and point
// `embeddingModelPath` at it. All matching happens on-device — no face
// image or embedding ever leaves the phone (see
// packages/core/src/driverProfile.ts for the storage/matching contract).

import android.graphics.Bitmap
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions

class DriverIdentifier(
    private val onFaceEmbedding: (WritableMap) -> Unit
) {
    // ACCURATE mode + landmarks so the crop fed to the embedding model is
    // well-aligned (eyes/nose/mouth), which matters a lot for embedding
    // quality with ResNet/FaceNet-style models.
    private val detector = FaceDetection.getClient(
        FaceDetectorOptions.Builder()
            .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_ACCURATE)
            .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_ALL)
            .setMinFaceSize(0.25f)
            .build()
    )

    // TODO: load the converted embedding model (see class doc above), e.g.
    // via TFLite Interpreter(FileUtil.loadMappedFile(context, embeddingModelPath))
    var embeddingModelPath: String? = null

    fun processFrame(bitmap: Bitmap, timestampMs: Long) {
        val image = InputImage.fromBitmap(bitmap, 0)
        detector.process(image)
            .addOnSuccessListener { faces ->
                val largest = faces.maxByOrNull { it.boundingBox.width() * it.boundingBox.height() }
                    ?: return@addOnSuccessListener

                // TODO: crop `largest.boundingBox` from bitmap, align via
                // eye landmarks, run through the embedding model, then emit.
                // val embedding = runEmbeddingModel(alignedCrop)
                // emit(embedding, timestampMs)
            }
    }

    private fun emit(embedding: FloatArray, timestampMs: Long) {
        val map: WritableMap = Arguments.createMap()
        val arr = Arguments.createArray()
        for (v in embedding) arr.pushDouble(v.toDouble())
        map.putArray("embedding", arr)
        map.putDouble("timestampMs", timestampMs.toDouble())
        onFaceEmbedding(map)
    }

    fun release() {
        detector.close()
    }
}
